File: Java/pattern/IteratorProj/src/com/pattern/MyIterator.java
package com.pattern;

public class MyIterator {
	private int index = 0;
	private Aggregation agg;
	private int size = 0;
	
	public MyIterator(Aggregation agg) {
		this.agg = agg;
		this.size = agg.getSize();
	}
	
	public void first() {
		index = 0;
	}
	
	public void next() {
		++index;
	}
	
	public boolean isDone() {
		if (index <= size -1 )
			return false;
		else
			return true;
	}
	
	public MyObject getCurrentElement() {
		return agg.getElement(index);
	}

}

File: cplusplus/design_pattern/IteratorProj/bak/MyIterator.cpp
#include "stdafx.h"
#include "MyIterator.h"


MyIterator::MyIterator()
{
    index = 0;
    size = 0;
}


MyIterator::~MyIterator()
{
}

MyIterator::MyIterator(const Aggregation & agg):agg(agg)
{
}

void MyIterator::first()
{
    index = 0;
}


void MyIterator::next()
{
    ++index;
}


bool MyIterator::isDone()
{
    return (index <= size) ? true : false;
}


std::shared_ptr<MyObject> MyIterator::getCurrentElement()
{
    return agg.getElement(index);
}
